File: Source/Raw.Net/Source/RawSpeedCode/CiffIFD.cs
#ifndef CIFF_IFD_H
#define CIFF_IFD_H

#include "FileMap.h"
#include "CiffEntry.h"
#include "CiffParserException.h"

/* 
    RawSpeed - RAW file decoder.

    Copyright (C) 2009-2014 Klaus Post
    Copyright (C) 2014 Pedro Côrte-Real

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

    http://www.klauspost.com
*/

namespace RawSpeed {

#define CIFF_DEPTH(_depth) if((depth=_depth+1) > 10) ThrowCPE("CIFF: sub-micron matryoshka dolls are ignored");

class CiffIFD
{
public:
  CiffIFD(FileMap* f, UInt32 start, UInt32 end, UInt32 depth=0);
  virtual ~CiffIFD(void);
  vector<CiffIFD*> mSubIFD;
  map<CiffTag, CiffEntry*> mEntry;
  vector<CiffIFD*> getIFDsWithTag(CiffTag tag);
  CiffEntry* getEntry(CiffTag tag);
  bool hasEntry(CiffTag tag);
  bool hasEntryRecursive(CiffTag tag);
  CiffEntry* getEntryRecursive(CiffTag tag);
  CiffEntry* getEntryRecursiveWhere(CiffTag tag, UInt32 isValue);
  CiffEntry* getEntryRecursiveWhere(CiffTag tag, string isValue);
  vector<CiffIFD*> getIFDsWithTagWhere(CiffTag tag, string isValue);
  vector<CiffIFD*> getIFDsWithTagWhere(CiffTag tag, UInt32 isValue);
  FileMap* getFileMap() {return mFile;};
protected:
  FileMap *mFile;
  UInt32 depth;
};

} // namespace RawSpeed

#endif
#include "StdAfx.h"
#include "CiffIFD.h"
#include "CiffParser.h"
  /*
  RawSpeed - RAW file decoder.

  Copyright (C) 2009-2014 Klaus Post
  Copyright (C) 2014 Pedro Côrte-Real

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

  http://www.klauspost.com
  */

namespace RawSpeed {

	CiffIFD::CiffIFD(FileMap* f, UInt32 start, UInt32 end, UInt32 _depth) {
		CIFF_DEPTH(_depth);
		mFile = f;

		UInt32 valuedata_size = get4LE(f.getData(end - 4, 4), 0);
		UInt16 dircount = get2LE(f.getData(start + valuedata_size, 2), 0);

		//  fprintf(stderr, "Found %d entries between %d and %d after %d data bytes\n", 
		//                  dircount, start, end, valuedata_size);

		for (UInt32 i = 0; i < dircount; i++) {
			int entry_offset = start + valuedata_size + 2 + i * 10;

			// If the space for the entry is no longer valid stop reading any more as
			// the file is broken or truncated
			if (!mFile.isValid(entry_offset, 10))
				break;

			CiffEntry *t = null;
			try {
				t = new CiffEntry(f, start, entry_offset);
			}
			catch (IOException) { // Ignore unparsable entry
				continue;
			}

			if (t.type == CIFF_SUB1 || t.type == CIFF_SUB2) {
				try {
					mSubIFD.push_back(new CiffIFD(f, t.data_offset, t.data_offset + t.bytesize, depth));
					delete(t);
				}
				catch (CiffParserException) { // Unparsable subifds are added as entries
					mEntry[t.tag] = t;
				}
				catch (IOException) { // Unparsable private data are added as entries
					mEntry[t.tag] = t;
				}
			}
			else {
				mEntry[t.tag] = t;
			}
		}
	}


	CiffIFD::~CiffIFD(void) {
		for (map<CiffTag, CiffEntry*>::iterator i = mEntry.begin(); i != mEntry.end(); ++i) {
			delete((*i).second);
		}
		mEntry.clear();
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			delete(*i);
		}
		mSubIFD.clear();
	}

	bool CiffIFD::hasEntryRecursive(CiffTag tag) {
		if (mEntry.find(tag) != mEntry.end())
			return true;
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			if ((*i).hasEntryRecursive(tag))
				return true;
		}
		return false;
	}

	vector<CiffIFD*> CiffIFD::getIFDsWithTag(CiffTag tag) {
		vector<CiffIFD*> matchingIFDs;
		if (mEntry.find(tag) != mEntry.end()) {
			matchingIFDs.push_back(this);
		}
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			vector<CiffIFD*> t = (*i).getIFDsWithTag(tag);
			for (UInt32 j = 0; j < t.size(); j++) {
				matchingIFDs.push_back(t[j]);
			}
		}
		return matchingIFDs;
	}

	vector<CiffIFD*> CiffIFD::getIFDsWithTagWhere(CiffTag tag, UInt32 isValue) {
		vector<CiffIFD*> matchingIFDs;
		if (mEntry.find(tag) != mEntry.end()) {
			CiffEntry* entry = mEntry[tag];
			if (entry.isInt() && entry.getInt() == isValue)
				matchingIFDs.push_back(this);
		}
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			vector<CiffIFD*> t = (*i).getIFDsWithTag(tag);
			for (UInt32 j = 0; j < t.size(); j++) {
				matchingIFDs.push_back(t[j]);
			}
		}
		return matchingIFDs;
	}

	vector<CiffIFD*> CiffIFD::getIFDsWithTagWhere(CiffTag tag, string isValue) {
		vector<CiffIFD*> matchingIFDs;
		if (mEntry.find(tag) != mEntry.end()) {
			CiffEntry* entry = mEntry[tag];
			if (entry.isString() && 0 == entry.getString().compare(isValue))
				matchingIFDs.push_back(this);
		}
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			vector<CiffIFD*> t = (*i).getIFDsWithTag(tag);
			for (UInt32 j = 0; j < t.size(); j++) {
				matchingIFDs.push_back(t[j]);
			}
		}
		return matchingIFDs;
	}

	CiffEntry* CiffIFD::getEntryRecursive(CiffTag tag) {
		if (mEntry.find(tag) != mEntry.end()) {
			return mEntry[tag];
		}
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			CiffEntry* entry = (*i).getEntryRecursive(tag);
			if (entry)
				return entry;
		}
		return null;
	}

	CiffEntry* CiffIFD::getEntryRecursiveWhere(CiffTag tag, UInt32 isValue) {
		if (mEntry.find(tag) != mEntry.end()) {
			CiffEntry* entry = mEntry[tag];
			if (entry.isInt() && entry.getInt() == isValue)
				return entry;
		}
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			CiffEntry* entry = (*i).getEntryRecursive(tag);
			if (entry)
				return entry;
		}
		return null;
	}

	CiffEntry* CiffIFD::getEntryRecursiveWhere(CiffTag tag, string isValue) {
		if (mEntry.find(tag) != mEntry.end()) {
			CiffEntry* entry = mEntry[tag];
			if (entry.isString() && 0 == entry.getString().compare(isValue))
				return entry;
		}
		for (vector<CiffIFD*>::iterator i = mSubIFD.begin(); i != mSubIFD.end(); ++i) {
			CiffEntry* entry = (*i).getEntryRecursive(tag);
			if (entry)
				return entry;
		}
		return null;
	}

	CiffEntry* CiffIFD::getEntry(CiffTag tag) {
		if (mEntry.find(tag) != mEntry.end()) {
			return mEntry[tag];
		}
		ThrowCPE("CiffIFD: CIFF Parser entry 0x%x not found.", tag);
		return 0;
	}


	bool CiffIFD::hasEntry(CiffTag tag) {
		return mEntry.find(tag) != mEntry.end();
	}

} // namespace RawSpeed

File: Source/RawSpeed/tiff/CiffIFD.cpp
/*
    RawSpeed - RAW file decoder.

    Copyright (C) 2009-2014 Klaus Post
    Copyright (C) 2014 Pedro Côrte-Real
    Copyright (C) 2017 Roman Lebedev

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include "tiff/CiffIFD.h"
#include "common/Common.h"               // for uint32, ushort16
#include "common/NORangesSet.h"          // for NORangesSet
#include "common/RawspeedException.h"    // for RawspeedException
#include "io/ByteStream.h"               // for ByteStream
#include "io/IOException.h"              // for IOException
#include "parsers/CiffParserException.h" // for ThrowCPE, CiffParserException
#include "tiff/CiffEntry.h"              // for CiffEntry, CiffDataType::CI...
#include <cassert>                       // for assert
#include <map>                           // for map, _Rb_tree_iterator
#include <memory>                        // for unique_ptr
#include <string>                        // for allocator, operator==, string
#include <utility>                       // for pair
#include <vector>                        // for vector

using std::string;
using std::vector;
using std::unique_ptr;

namespace rawspeed {

void CiffIFD::parseIFDEntry(NORangesSet<Buffer>* ifds, ByteStream* bs) {
  assert(ifds);
  assert(bs);

  unique_ptr<CiffEntry> t;

  auto origPos = bs->getPosition();

  try {
    t = std::make_unique<CiffEntry>(bs);
  } catch (IOException&) {
    // Ignore unparsable entry, but fix probably broken position due to
    // interruption by exception; i.e. setting it to the next entry.
    bs->setPosition(origPos + 10);
    return;
  }

  try {
    switch (t->type) {
    case CIFF_SUB1:
    case CIFF_SUB2: {
      add(std::make_unique<CiffIFD>(this, ifds, &t->data));
      break;
    }

    default:
      add(move(t));
    }
  } catch (RawspeedException&) {
    // Unparsable private data are added as entries
    add(move(t));
  }
}

CiffIFD::CiffIFD(const CiffIFD* parent_, NORangesSet<Buffer>* ifds,
                 ByteStream* mFile)
    : parent(parent_) {
  assert(ifds);
  assert(mFile);

  checkOverflow();

  if (mFile->getSize() < 4)
    ThrowCPE("File is probably corrupted.");

  // last 4 bytes is the offset to the beginning of the [first?] IFD
  mFile->setPosition(mFile->getSize() - 4);
  uint32 offset = mFile->getU32();
  mFile->setPosition(offset);

  // count of the Directory entries in this IFD
  ushort16 dircount = mFile->getU16();

  // 2 bytes for entry count
  // each entry is 10 bytes
  const auto IFDFullSize = 2 + 10 * dircount;
  const Buffer IFDBuf(mFile->getSubView(offset, IFDFullSize));
  if (!ifds->emplace(IFDBuf).second)
    ThrowCPE("Two IFD's overlap. Raw corrupt!");

  for (uint32 i = 0; i < dircount; i++)
    parseIFDEntry(ifds, mFile);
}

void CiffIFD::checkOverflow() const {
  const CiffIFD* p = this;
  int i = 0;
  while ((p = p->parent) != nullptr) {
    i++;
    if (i > 5)
      ThrowCPE("CiffIFD cascading overflow.");
  }
}

void CiffIFD::add(std::unique_ptr<CiffIFD> subIFD) {
  if (mSubIFD.size() > 100)
    ThrowCPE("CIFF file has too many SubIFDs, probably broken");

  subIFD->parent = this;
  mSubIFD.push_back(move(subIFD));
}

void CiffIFD::add(std::unique_ptr<CiffEntry> entry) {
  entry->parent = this;
  mEntry[entry->tag] = move(entry);
}

template <typename Lambda>
std::vector<const CiffIFD*> CiffIFD::getIFDsWithTagIf(CiffTag tag,
                                                      const Lambda& f) const {
  std::vector<const CiffIFD*> matchingIFDs;

  const auto found = mEntry.find(tag);
  if (found != mEntry.end()) {
    const auto entry = found->second.get();
    if (f(entry))
      matchingIFDs.push_back(this);
  }

  for (const auto& i : mSubIFD) {
    const auto t = i->getIFDsWithTagIf(tag, f);
    matchingIFDs.insert(matchingIFDs.end(), t.begin(), t.end());
  }

  return matchingIFDs;
}

template <typename Lambda>
const CiffEntry* CiffIFD::getEntryRecursiveIf(CiffTag tag,
                                              const Lambda& f) const {
  const auto found = mEntry.find(tag);
  if (found != mEntry.end()) {
    const auto entry = found->second.get();
    if (f(entry))
      return entry;
  }

  for (const auto& i : mSubIFD) {
    const CiffEntry* entry = i->getEntryRecursiveIf(tag, f);
    if (entry)
      return entry;
  }

  return nullptr;
}

vector<const CiffIFD*> CiffIFD::getIFDsWithTag(CiffTag tag) const {
  return getIFDsWithTagIf(tag, [](const CiffEntry*) { return true; });
}

vector<const CiffIFD*> CiffIFD::getIFDsWithTagWhere(CiffTag tag,
                                                    uint32 isValue) const {
  return getIFDsWithTagIf(tag, [&isValue](const CiffEntry* entry) {
    return entry->isInt() && entry->getU32() == isValue;
  });
}

vector<const CiffIFD*>
CiffIFD::getIFDsWithTagWhere(CiffTag tag, const string& isValue) const {
  return getIFDsWithTagIf(tag, [&isValue](const CiffEntry* entry) {
    return entry->isString() && isValue == entry->getString();
  });
}

bool __attribute__((pure)) CiffIFD::hasEntry(CiffTag tag) const {
  return mEntry.count(tag) > 0;
}

bool __attribute__((pure)) CiffIFD::hasEntryRecursive(CiffTag tag) const {
  if (mEntry.count(tag) > 0)
    return true;

  for (const auto& i : mSubIFD) {
    if (i->hasEntryRecursive(tag))
      return true;
  }

  return false;
}

const CiffEntry* CiffIFD::getEntry(CiffTag tag) const {
  const auto found = mEntry.find(tag);
  if (found != mEntry.end())
    return found->second.get();

  ThrowCPE("Entry 0x%x not found.", tag);
}

const CiffEntry* CiffIFD::getEntryRecursive(CiffTag tag) const {
  return getEntryRecursiveIf(tag, [](const CiffEntry*) { return true; });
}

const CiffEntry* CiffIFD::getEntryRecursiveWhere(CiffTag tag,
                                                 uint32 isValue) const {
  return getEntryRecursiveIf(tag, [&isValue](const CiffEntry* entry) {
    return entry->isInt() && entry->getU32() == isValue;
  });
}

const CiffEntry* CiffIFD::getEntryRecursiveWhere(CiffTag tag,
                                                 const string& isValue) const {
  return getEntryRecursiveIf(tag, [&isValue](const CiffEntry* entry) {
    return entry->isString() && isValue == entry->getString();
  });
}

} // namespace rawspeed
